pkg/networkserver/mac/link_adr.go

```diff
@@ -321,7 +322,10 @@ func HandleLinkADRAns(
 	}
 
 	ev := EvtReceiveLinkADRAccept
-	if !pld.ChannelMaskAck || !pld.DataRateIndexAck || !pld.TxPowerIndexAck {
+
+	if (!adrEnabled && !pld.ChannelMaskAck) ||
+		(adrEnabled && !pld.DataRateIndexAck) ||
+		(adrEnabled && !pld.TxPowerIndexAck) {
```

Comment:
What is the relevant spec again? Isn't it the same rule for all acks? Shouldn't this be so that if ADR is not enabled, that any of the bits can be not-acked, while if ADR is enabled, all bits must be acked?

Can you reference the spec section here on L326 in a comment?

Also, can you check if the behavior changed since 1.0.2, 1.0.3 and 1.0.4?

Reply:
> What is the relevant spec again? Isn't it the same rule for all acks? Shouldn't this be so that if ADR is not enabled, that any of the bits can be not-acked, while if ADR is enabled, all bits must be acked?
> 
LoRaWAN 1.0.4 spec L534-538:
```
An end-device SHOULD accept the channel mask controls present in LinkADRReq, even
when the ADR bit is not set. The end-device SHALL respond to all LinkADRReq commands
with a LinkADRAns indicating which command elements were accepted and which were
rejected. This behavior differs from when the uplink ADR bit is set, in which case the end-
device accepts or rejects the entire command.
```

So if the ADR bit is set then the end device should accept the entire command, means any of the ACK bit is not set -> rejected.
If the ADR bit is not set then the end device should accept the channel mask, means if the `ChannelMaskAck` bit is not set then is rejected, the `DataRateIndexAck` and `TxPowerIndexAck` doesn't matter.

> Can you reference the spec section here on L326 in a comment?
> 
Yes, will do.

> Also, can you check if the behavior changed since 1.0.2, 1.0.3 and 1.0.4?
When I first checked the bug report then I only checked the 1.0.4 (my bad, sorry) but now checked all the version (1.0.2, 1.0.3, 1.0.4, 1.1) and it turns out it is a new change, only the 1.0.4 describes this behaviour (1.1 also doesn't have it) and 1.0.3 has an errata that describes this change https://resources.lora-alliance.org/home/adr-harmonize-state-transition-errata-on-the-lorawan-l2-1-0-3-specification So I'm assuming 1.0.3 also has this behaviour. -> I need to check the lorawan version in this logic.